### PR TITLE
[shortfin-sd] Add exports and support for scheduled unet, batch sizes.

### PR DESCRIPTION
**Pull request by monorimet**

### CODE REVIEW

**Comment by monorimet** [comment]
image
turbine_generate action works really nicely, with multiple batch size export and compile jobs running in parallel!

**Review by daveliddell — APPROVED** [review]
LGTM! I can't say that I understand everything, but of what I do understand, I can't find anything wrong with it.  :-)

File: sharktank/sharktank/models/punet/model.py [hunk 1]
```diff
@@ -147,7 +147,7 @@ def forward(
         # 1b. Aug embedding of text_embeds, time_ids
         time_embeds = self.add_time_proj(time_ids.flatten())
         time_embeds = time_embeds.reshape((text_embeds.shape[0], -1))
-        add_embeds = torch.concat([text_embeds, time_embeds], dim=-1).to(emb.dtype)
+        add_embeds = torch.cat([text_embeds, time_embeds], dim=-1).to(emb.dtype)
```
[the remark this answers is not in the document]
Comment (monorimet) [reply sharktank/sharktank/models/punet/model.py:150]: Yes. Beyond torch 2.5, torch.concat no longer exports to the same thing as torch.cat, and the former runs into compiler issues that don't happen with the latter.

File: shortfin/python/shortfin_apps/sd/components/exports.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    if decomp_attn == True:
+        decomp_list = [
+            torch.ops.aten._scaled_dot_product_flash_attention_for_cpu,
+            torch.ops.aten._scaled_dot_product_flash_attention.default,
+            torch.ops.aten.scaled_dot_product_attention.default,
+            torch.ops.aten.scaled_dot_product_attention,
+        ]
+    with decompositions.extend_aot_decompositions(
+        from_current=True,
+        add_ops=decomp_list,
```
[the remark this answers is not in the document]
Comment (monorimet) [reply shortfin/python/shortfin_apps/sd/components/exports.py:41]: It only does this if the decomp_list contains sdpa, which is guarded by the decomp_attn argument.

File: sharktank/sharktank/torch_exports/sdxl/clip.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    def __init__(
+        self,
+        hf_model_name,
+        precision,
+        batch_size=1,
+        batch_input=False,
+    ):
+        super().__init__()
+        self.torch_dtype = torch.float16 if precision == "fp16" else torch.float32
+        config_1 = CLIPTextConfig.from_pretrained(
```
[the remark this answers is not in the document]
Comment (monorimet) [reply sharktank/sharktank/torch_exports/sdxl/clip.py:35]: Yeah. At the time I didn't want to introduce another model source discrepancy between this and turbine-models, but definitely want to have sharktank exports plugged in where possible.

File: sharktank/sharktank/torch_exports/sdxl/scheduler.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+        self.model = scheduler
+        self.height = height
+        self.width = width
+        self.is_sd3 = False
+        if "stable-diffusion-3" in hf_model_name:
+            self.is_sd3 = True
+        self.batch_size = batch_size
+        # Whether this will be used with CFG-enabled pipeline.
+        self.do_classifier_free_guidance = True
+        timesteps = [torch.empty((100), dtype=dtype, requires_grad=False)] * 100
```
Comment (IanNod) [inline sharktank/sharktank/torch_exports/sdxl/scheduler.py:36]: would be good to have a comment for what the hardcoded 100 is

File: shortfin/python/shortfin_apps/sd/components/builders.py [hunk 5]
```diff
@@ -83,57 +107,217 @@ def get_params_filenames(model_params: ModelParams, model=None, splat: bool = Fa
...
+        self.executor.write_status(f"Fetching URL: {self.url} -> {path}")
+        try:
+            urllib.request.urlretrieve(self.url, str(path))
+        except urllib.error.HTTPError as e:
+            if retries > 0:
+                retries -= 1
+                self._invoke(retries=retries)
+            else:
+                raise IOError(f"Failed to fetch URL '{self.url}': {e}") from None
+        local_size = get_file_size(str(path))
```
Comment (IanNod) [inline shortfin/python/shortfin_apps/sd/components/builders.py:286]: comparing and md5 hash would be more robust

File: shortfin/python/shortfin_apps/sd/components/exports.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (IanNod) [reply shortfin/python/shortfin_apps/sd/components/exports.py:41]: ah, misread this originally as a single function to export all models, not called individually for each component.

**Review by IanNod — APPROVED** [review]

File: sharktank/sharktank/models/punet/model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (IanNod) [reply sharktank/sharktank/models/punet/model.py:150]: Interesting, torch.concat is supposed to be just an alias of torch.cat. I guess the docs lie https://pytorch.org/docs/stable/generated/torch.concat.html

File: sharktank/sharktank/torch_exports/sdxl/unet.py [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+            "config.json": download("config.json"),
+        }
+    if quant_paths and quant_paths["params"] and os.path.exists(quant_paths["params"]):
+        results["params.safetensors"] = quant_paths["params"]
+    else:
+        results["params.safetensors"] = download("params.safetensors")
+
+    output_dir = os.path.dirname(external_weight_path)
+
+    # if os.path.exists(external_weight_path):
```
Comment (IanNod) [inline sharktank/sharktank/torch_exports/sdxl/unet.py:297]: commented code?

**Comment by monorimet** [comment]
Reopened as closing was not intended.

File: shortfin/python/shortfin_apps/sd/components/builders.py (hunk 5 above)
Comment (monorimet) [reply shortfin/python/shortfin_apps/sd/components/builders.py:284]: Yeah this is supposed to help if a user's download of parameters gets interrupted somehow -- md5sum was skipped here as it may some time to generate for large files.
